### PR TITLE
Many major changes.

### PR DESCRIPTION
**Pull request by mpellicer**
This PR includes many major changes.

- Makes the logic clearer.
- Methods involving fetches are now asynchronous to avoid the callback hell, providing better structure.
- Nested logic has been moved into the main JS thread.
- All the logic has been structured and moved into separate and smaller functions.
- Global variables cleanup, now used in the functions.
- Conditionals extracted out of the functions.
- Utility methods extracted into separate functions.
- Commented logic cleanup, all the logic is in previous commits so nothing is lost.

The refactor is not 100% completed as many logic can be refactored too.

The legacy bug relared to the progress bar and elements not loaded yet has been resolved, now it does not depend on timeouts, this has been tested simulating slow connections in Chrome and works fine.

I've created this PR because involves many changes and the review process can be long.

### CODE REVIEW

**Comment by mpellicer** [comment]
In first place I used that method but is never triggered, at least in Chrome, waiting for the load just fixes a bug rendering the progress bar.

**Comment by buckett** [comment]
But that can be a very long time, especially if some of the other JS/CSS/images on the page are taking a while to load.

**Comment by buckett** [comment]
The DOMContentLoaded event doesn't fire for handlers that are registered after it has passed.

**Comment by mpellicer** [comment]
Yeah, makes sense and it's kinda bad, the DOMContentLoaded event sounds optimal to me.

This is just to fix the progress bar bug, sometimes gets the footer and sometimes not, depends on how fast it loads, it also uses a different id when you're in the last page of a module.

**Comment by mpellicer** [comment]
> But that can be a very long time, especially if some of the other JS/CSS/images on the page are taking a while to load.

Is that a problem? I tested with slow connections (Fast 3G) and the UI behavior is not very different than the existing behavior.

**Comment by buckett** [comment]
Yeah, because we want the CPN changes to apply to the DOM as early as possible so you don't get flicker later on.

**Comment by mpellicer** [comment]
Performing the script after the load event produces the best behavior around the progress bar and works in all the cases and using slow connections. The current behavior (using setTimeout) is not consistent, the progress bar appears sometimes, depending if the DIV elements were added to the DOM after or before executing that portion of the script.

The flicker happens always, the modules are always rendered in Canvas before the script.

**Comment by buckett** [comment]
My problem with the load event is that if you have a image like this in your home page: 
https://deelay.me/5000/https://picsum.photos/200/300 

Then the load event doesn't fire for 5 seconds. Waiting this long means that the left menu jumps about long after 99% of the page has loaded.

If you are waiting for elements to appear in the page then using a multation observer is a nice way to watch for them:
https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver
Does this allow you to add the progress bar only when the required elements are loaded?

File: canvas-where-am-I.js [hunk 1]
```diff
@@ -79,524 +96,283 @@
...
+
+        let moduleTileContent = document.createElement('div');
+        moduleTileContent.className = 'ou-ModuleCard__header_content';
+
+        var moduleTileTitle = document.createElement('div');
+        moduleTileTitle.classList.add('ou-ModuleCard__header-title');
+        moduleTileTitle.classList.add('ellipsis');
+        moduleTileTitle.title = module.name;
+        moduleTileTitle.style.color = moduleColours[mindex];
+        moduleTileTitle.innerHTML = module.name;
```
Comment (buckett) [inline canvas-where-am-I.js:206]: I don't think we want to do this as now if someone has a script tag in their module name we have just injected it into the page. My first attempt at preventing things would be to add a text node to the div and then set the name on the text node (but I haven't tested this).

File: canvas-where-am-I.js [hunk 2]
```diff
@@ -79,524 +96,283 @@
...
+        // Create a new item for the submodule list.
+        let newItem = document.createElement('li');
+        newItem.className = 'ou-section-sub';
+
+        // Create a new Link for the submodule item.
+        let newLink = document.createElement('a');
+        newLink.className = 'ou-section-link-sub';
+        newLink.title = module.name;
+        newLink.href = `/courses/${courseId}/modules/${module.id}`;
+        newLink.innerHTML = module.name;
```
Comment (buckett) [inline canvas-where-am-I.js:245]: I don't think we want to do this as now if someone has a script tag in their module name we have just injected it into the page. My first attempt at preventing things would be to add a text node to the div and then set the name on the text node (but I haven't tested this).

File: canvas-where-am-I.js [hunk 3]
```diff
@@ -648,47 +424,52 @@
...
+              return 'icon-discussion';
+          case 'Quiz':
+              return 'icon-quiz';
+          case 'Assignment':
+              return 'icon-assignment';
+          case 'ExternalUrl':
+              return 'icon-link';
+          default:
+              return'icon-document';
+      }
```
Comment (buckett) [inline canvas-where-am-I.js:468]: 👍

**Review by buckett — APPROVED** [review]
Nice work, I know you didn't introduce the innerHtml problem, but I'm keen we address it and don't accidentally forget about it.

**Comment by mpellicer** [comment]
> My problem with the load event is that if you have a image like this in your home page:
> https://deelay.me/5000/https://picsum.photos/200/300
> 
> Then the load event doesn't fire for 5 seconds. Waiting this long means that the left menu jumps about long after 99% of the page has loaded.
> 
> If you are waiting for elements to appear in the page then using a multation observer is a nice way to watch for them:
> https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver
> Does this allow you to add the progress bar only when the required elements are loaded?

Makes sense and you're right, pages with many assets can take ages to perform the CPN replacements,  I'll try to work on the mutation observer, catch the changes and perform our logic.

File: canvas-where-am-I.js [hunk 4]
```diff
@@ -648,47 +424,52 @@
...
-        if (startPos != -1) {
-            startPos = startPos + moduleIdTerm.length; //account for length of moduleItemTerm as found beginning position
-            moduleId = parseInt(currentUrl.slice(startPos)); //will substring from end
-        }
-        return moduleId;
+        const moduleHash = window.location.hash.substr(1);
+        const moduleHashPrefix = 'module_';
+        // If the module hash starts with the module_ prefix, remove the prefix to get the Id.
+        // Otherwise return 0, moduleId not found
+        return moduleHash.startsWith(moduleHashPrefix) ? moduleHash.replace(moduleHashPrefix, '') : 0;
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply canvas-where-am-I.js:446]: Sounds good, return this as integer.

File: canvas-where-am-I.js (hunk 1 above)
Comment (buckett) [reply canvas-where-am-I.js:206]: An example of a bad title is: `<span onclick="alert('hello')">Text</span>`

File: canvas-where-am-I.js (hunk 2 above)
Comment (mpellicer) [reply canvas-where-am-I.js]: Sure, let me change the logic.

File: canvas-where-am-I.js [hunk 5]
```diff
@@ -648,47 +424,52 @@
...
-            //for external links at least, we have a URL in the format: /courses/13199/modules/items/87888 so let's seee if we can extract from that
-            moduleItemTerm = '/modules/items/';
-            startPos = currentUrl.indexOf(moduleItemTerm); //is this in URL
-            if (startPos != -1) {
-                startPos = startPos + moduleItemTerm.length; //account for length of moduleItemTerm as found beginning position
-                moduleItemId = parseInt(currentUrl.slice(startPos)); //will substring from end
-            }
-        }
-        return moduleItemId;
+        const moduleRequestUrl = new URL(window.location.href);
```
[the remark this answers is not in the document]
Comment (mpellicer) [reply canvas-where-am-I.js:429]: Canvas does not support IE11 and doesn't work anymore, I tested it and blocks the UI displaying this message so this logic will never be performed by IE11.
image

**Comment by mpellicer** [comment]
The observer works similar than the timeout approach but worse than waiting for the load event, there's still one edge case when the observer is defined after the DOM performs the changes, not cool :(

**Comment by mpellicer** [comment]
This latest commit works fine, waits until gets the element and then renders the progress bar:
https://github.com/oxctl/canvas-where-am-I/pull/3/commits/3fba8d99d378a39306100664c8568ad221c14210

I also included a maximum number of retries for the case that the id does not exist, tested and aborts correctly.

**Pull request merged**